### ISSUE TITLE
agent can benefit from keeping intended Transactions not coupled into intended Exchanges

### ISSUE THREAD

**Issue by elf-pavlik** [valueflows/intent#5]
I find this point quite important and capturing it well might take me few iterations. Let's take example where an agent "Bumblebee Coop" might among many intents cast two specific ones:
1) Sell 50kg of apples for 1000MXN
2) Pay for transporting 500kg of apples 500MXN

Since we don't give monetary currencies a privileged role. We model this Offer and Request as two pairs of two intended vf:Exchange, each composed of two intended vf:Transfers.

?foo - means variable

### Offer
* vf:Transfer ( from: "Bumblebee Coop", to: ?consumer, rightsTo: "50kg of apples")
* vf:Transfer ( from: ?consumer, to: "Bumblebee Coop", rightsTo: "1000MXN" )

### Request
* vf:Transfer ( from: ?provider, to: "Bumblebee Coop", rightsTo: "Transport Service for 500kg of apples")
* vf:Transfer ( from: "Bumblebee Coop", to: ?provider, rightsTo: "500MXN" )

Above pairing of intended transfers I would consider "tightly coupled" (from current lack of better term)

If we define those intended transfers as more independent we would heave

### Intent
* vf:Transfer ( from: "Bumblebee Coop", to: ?consumer, rightsTo: "50kg of apples")
* vf:Transfer ( from: ?provider, to: "Bumblebee Coop", rightsTo: "Transport Service for 500kg of apples")

(let's leave playing with the MXN aside for now)

With such independent intents, another agent - "Funky Wheelers", might **propose** vf:Exchange:

### Exchange (proposed)
* vf:Transfer ( from: "Funky Wheelers", to: "Bumblebee Coop", rightsTo: "Transport Service for 500kg of apples")
* vf:Transfer ( from: "Bumblebee Coop", to: "Funky Wheelers", rightsTo: "25kg of apples")

above might not happen if agents kept their **intended transfers** tightly coupled into 'intended exchanges'...

**Comment by fosterlynn** [comment]
Makes sense to me @elf-pavlik .  I can see people wanting to do it either way though.  Some people may have a need to keep the tight reciprocity as part of their Intent.  Others may be happy to leave it loose and see what comes up.

I think of people advertising in our local email listing, some say they want $10 for something they are offering, some say they will pay or barter for something they want, sometimes specific barter offer, sometimes leave it open.  It's a mix.

**Comment by elf-pavlik** [comment]
This issue needs update to incorporate recent findings on building exchanges around events instead of focusing on resources, https://github.com/valueflows/exchange/issues/34

**Comment by fosterlynn** [comment]
Agreed, we need to do some more detailed work on Intents, maybe this is a good place to do it...

**Comment by elf-pavlik** [comment]
For now in examples here I mostly want to replace 'transfer of rights' with 'making commitments'

> vf:Transfer ( from: ?provider, to: "Bumblebee Coop", rightsTo: "Transport Service for 500kg of apples")

This would change to intents for *load* & *unload*

What I find even more interesting and which goes beyond this issue. Looking at https://github.com/valueflows/valueflows/blob/master/use-cases/crowdfunding-in-kind.md#in-kind--monetary-hybrid If Bumblebee Coop had intents to exchange Apples for MEX just because they needed MEX to exchange fore transporting apples (we can replace here MEX with Faircoin or some specific credits issued by transportation provider who Bumblebee currently receive services from). Making an agreement for transporting apples in exchange for apples, would result in canceling the intent to exchange apples for MEX (or Faircoin or those specific credits from transportation provider)

**Comment by fosterlynn** [comment]
>For now in examples here I mostly want to replace 'transfer of rights' with 'making commitments'

Yes, 'transfer of rights' is out of date I guess. :blush: But would the vf:action go there?

I have been wondering if we can get Intents to work with vf:action, or if that will be too confusing for people.  I'll do some yaml when I have a block of time to think about it.

**Comment by fosterlynn** [comment]
And part of it too is how to represent the tight or loose coupling of the offer with what you want in reciprocity for the offer...

**Comment by elf-pavlik** [comment]
> And part of it too is how to represent the tight or loose coupling of the offer with what you want in reciprocity for the offer...

If we call
* Intent to act as *provider* - offer
* Intent to act as *receiver* - request

I see need for having a way to have a way to express 'aggregated' intents (offers/requests) with quantities for certain period of time
* offer - act as *provider* on *issue* apples - 3000kg in next 30 days
* request - act as *receiver* on *receive* peanuts - 200kg in next 30 days
* offer - act as *provider* on *issue* Faircoin - 2000 in next 30 days
* request - act as *receiver* on *receive* Faircoin - 800 in next 30 days

Where agent would do that based on all the planned processes. In addition to that, agent could publish 'suggested exchanges' which would pair certain sub-quantities of offers & requests at the same time having a way for chosen aggregated intents to provide way for other agents to suggest other exchanges

One could also make available events that *fulfill* those intents and commitments that should *fulfill* those intents as a way for others to calculate remaining quantities. In case agent doesn't want to expose those events and commitments we may need a way to reference calculated 'remaining quantity' directly from those intents.

**Comment by fosterlynn** [comment]
>I see need for having a way to express 'aggregated' intents (offers/requests) with quantities for certain period of time
>In addition to that, agent could publish 'suggested exchanges' 

Yup. :+1:

**Comment by elf-pavlik** [comment]
rough draft of YAML with stocks and *receive* / *issue* intents on those stocks

```yaml
'@context': https://w3id.org/valueflows/v1-dev
id: https://bumblebee.example/bd8c416c-8f93-420f-b945-153ddc3a5b9f#stock
type: Stock
category: https://www.wikidata.org/wiki/Q89 # apple
quantity:
  numericValue: 500
  unit: unit:Kilogram
  
'@context': https://w3id.org/valueflows/v1-dev
id: https://bumblebee.example/5fb45882-afe3-4eb5-a9c4-99a6f3a32bc4#stock
type: Stock
category: https://www.wikidata.org/wiki/Q3621199 # peanut
quantity:
  numericValue: 0
  unit: unit:Kilogram
  
'@context': https://w3id.org/valueflows/v1-dev
id: https://bumblebee.example/09369ce2-7fd2-4714-ad99-494b2e8d595f#stock
type: Stock
category: https://www.wikidata.org/entity/Q4730 #MXN
quantity:
  numericValue: 500
  unit: iso4217:MXN
  
'@context': https://w3id.org/valueflows/v1-dev
id: https://bumblebee.example/b471a8ba-992d-49ce-afd9-b99ed70c4cc2#intent
type: Intent
provider: https://bumblebee.exampe/#identity
validFrom: 2017-03-01
validUntil: 2017-03-31
action: issue
decrementQuantity: 
    numericValue: 3000
    unit: iso4217:MXN
involveResource: https://bumblebee.example/09369ce2-7fd2-4714-ad99-494b2e8d595f#stock # MXN
  
'@context': https://w3id.org/valueflows/v1-dev
id: https://bumblebee.example/a265fddf-0905-49e3-8c5b-a0d6e2c85386#intent
type: Intent
receiver: https://bumblebee.exampe/#identity
validFrom: 2017-03-01
validUntil: 2017-03-31
action: receive
incrementQuantity: 
    numericValue: 2500
    unit: iso4217:MXN
involveResource: https://bumblebee.example/09369ce2-7fd2-4714-ad99-494b2e8d595f#stock # MXN
  
'@context': https://w3id.org/valueflows/v1-dev
id: https://bumblebee.example/7178082f-f41c-4b17-b58a-0cd0931bf423#intent
type: Intent
provider: https://bumblebee.exampe/#identity
validFrom: 2017-03-01
validUntil: 2017-03-31
action: issue
decrementQuantity: 
    numericValue: 300
    unit: unit:Kilogram
involveResource: https://bumblebee.example/bd8c416c-8f93-420f-b945-153ddc3a5b9f#stock # apples

'@context': https://w3id.org/valueflows/v1-dev
id: https://bumblebee.example/9bf6a965-9885-4c28-8e05-973f835a5ac6#intent
type: Intent
receiver: https://bumblebee.exampe/#identity
validFrom: 2017-03-01
validUntil: 2017-03-31
action: receive
incrementQuantity: 
    numericValue: 30
    unit: unit:Kilogram
involveResource: https://bumblebee.example/5fb45882-afe3-4eb5-a9c4-99a6f3a32bc4#stock # peanuts
```

1) whenever Event happens which partially fulfills any of those intents, they should link back to it so everyone can calculate 'remaining quantities'
2) similar for Commitments made for those Intents - calculate 'not committed yet quantities'
3) in relation to #7 (matches) with more stocks of resources with all kind of categories, and all kind of currency resources, plans can include some options of 'intended exchanges', when alternative events or commitments appear for those it should affect exchanges. for example if someone had 'intended exchange': 50kg of apples for 1000MXN and 1000MXN for 10kg of peanuts, when commitment appears for exchange: 50kg of apples for 10kg of peanuts, which satisfies certain demand, it should replace original intents to issue/receive MXN with new ones, both with quantities decremented by 1000MXN - https://github.com/valueflows/valueflows/blob/master/use-cases/crowdfunding-in-kind.md#in-kind--monetary-hybrid

In short, plans on particular stocks would describe I/O related to exchanges *issue* & *receive*, currencies only serve purpose of facilitating some of the intended exchanges. When exchange happens or commitments get made, it should affect intents related to stocks of currencies considered as options to facilitate exchanges related to all the non currency stocks.

**Comment by fosterlynn** [comment]
>whenever Event happens which partially fulfills any of those intents, they should link back to it so everyone can calculate 'remaining quantities'

:+1: 

>similar for Commitments made for those Intents - calculate 'not committed yet quantities'

:+1: 

>in relation to #7 (matches) with more stocks of resources with all kind of categories, and all kind of currency resources, plans can include some options of 'intended exchanges',

I looked at #7... what do you think would be the best way at this point to hook together intents when one is what you are offering say, and one is what you would like in return?  I don't mind having them separate, but some use cases will need them associated.

Something like:

``` 
'@context': https://w3id.org/valueflows/v1-dev
id: https://bumblebee.example/7178082f-f41c-4b17-b58a-0cd0931bf423#intent
type: Intent
provider: https://bumblebee.exampe/#identity
validFrom: 2017-03-01
validUntil: 2017-03-31
action: issue
intentQuantity: 
    numericValue: 300
    unit: unit:Kilogram
involveResource: https://bumblebee.example/bd8c416c-8f93-420f-b945-153ddc3a5b9f#stock # apples
reciprocity:
    action: receive
    intentQuantity: 
        numericValue: 2500
        unit: iso4217:MXN
```

Or here's a less involved possible example, just brainstorming:

``` 
'@context': https://w3id.org/valueflows/v1-dev
id: https://bumblebee.example/7178082f-f41c-4b17-b58a-0cd0931bf423#intent
type: Intent
provider: https://bumblebee.exampe/#identity
validFrom: 2017-03-01
validUntil: 2017-03-31
action: issue
note: I am offering number 2 quality apples, Haralsons and Johnagolds mixed, 
         from last year's crop.  Still in great shape for this late in the season.
reciprocity:
    action: receive
    intentQuantity: 
        numericValue: 25
        unit: iso4217:MXN
    perQuantity:
        numericValue: 1
        unit: unit:Kilogram
```

**Comment by elf-pavlik** [comment]
I understand that you have 2 instances of intent in both snippets, and they just miss `@type` in two places

```yaml
reciprocity:
    type: Intent
    action: receive
```

Instead of having Intents referencing each other directly, I think we may better have them connected through Offer / Request. Similar as *Agreement* connects many Intents, Commitments and Events. We can see Offer / Request as suggestion of *Agreement* which through CfA becomes an Agreement - CfA as simple as yes/no or some complex negotiation using that Offer / Request just as a starting point.

**Comment by fosterlynn** [comment]
I'm not sure we want a whole intent as the reciprocal part.  One reason I think it can be different is that if we are using this for pricing, it will be for the minimum amount.  So I'm not sure it is really an Intent, more parts of an intent to let people know a starting place for CfA, or just the price.  For example it could just have a note saying "Will barter for any kind of fresh veggie or free range eggs, or just make an offer."

But you could be right, we can do it your way for now.  

If you wouldn't connect them directly, what would it look like?  I am fine with disconnected Intents for the use case where people are just posting stuff they want and offer, and there will be a lot of mix and matching.  But I don't ever see currency as a disconnected Intent except for crowdfunding or something like that.  And I think in most cases, there will be a direct connection, since there aren't many communities that are organized for disconnected barter.

**Comment by elf-pavlik** [comment]
I'll try to write some YAML today/tomorrow. What we should keep in mind: **multiple events can partially fulfill single intent**. So if agents have their *primary* intents for let's say upcoming months (as we discussed on gitter *primary* intents shouldn't include currencies). Those intents can get fulfilled in parts through events in each exchange made during that month. Possibly Offer-Request could reference at least one primary Intents and provide conditions on partial fulfillment for it. As we discussed in some cases of barter Offer-Request can reference primary Intent where agent making that Offer-Request appears as *provider* and another primary Intent where agent making it appears as *receiver*. Eg. *using my car for me using someone's bike* (two primary intents) Offer-Request would reference both and provide some 'ratio/rate' for exchanges in which events would partially fulfill both of those *primary* intents. Intents with currencies which we can consider *secondary* would appear as alternative options in those Offer-Request eg. *using my car for currency A or currency B or currency C* giving 'ratios/rates' one per primary intent and any of those options for secondary intents (currencies).

**Comment by fosterlynn** [comment]
Great, I'll look forward to the yaml, and we can clarify the names (Intent, Offer, Request).

>What we should keep in mind: multiple events can partially fulfill single intent.

Yes, for sure. 

>(as we discussed on gitter primary intents shouldn't include currencies)

That made sense to me at the time, when we were talking about offers of goods and services for currency.  But the exception I thought of is when actually requesting money for something specific, like crowdfunding.  Just not the "sure everyone can use more cash" kind of thing.  OK?

**Comment by elf-pavlik** [comment]
> But the exception I thought of is when actually requesting money for something specific, like crowdfunding. Just not the "sure everyone can use more cash" kind of thing. OK?

As explained in https://github.com/valueflows/valueflows/blob/master/use-cases/crowdfunding-in-kind.md#in-kind--monetary-hybrid even with crowdfunging intent to receive currency(ies) doesn't come as *primary*! Agent who raises funds in practice has some *primary* intents + knows matches which require some currencies in exchange that can lead to fulfillment of those *primary* intents. Any Commitments on those *primary* intents that get made not as part as exchange agreement which involves one of the currencies agent tries to 'crowdfund', but possibly as *in-kind* donation, should cause updating that *secondary* intent to *receive* currency(ies) with one that gets decremented by the amount from previously considered match that would fulfill that *primary* intent in a part equal to that Commitment which didn't required currency in exchange. I think I should illustrate it :smile:

**Comment by fosterlynn** [comment]
That's totally fine for crowdfunding-in-kind.  But that is your vision for how to do crowdfunding better, not the way people actually crowdfund most of the time....  or at least some of the time....  :)

**Comment by elf-pavlik** [comment]
Emphasis on in-kind or not don't change case of currency related intents not acting as *primary*. Someone who does a serious crowdfunding campaign, first would try to 'estimate expenses'. This estimate will rely an all the *primary* intents (non-currency) one. Then based on those *primary* intents one would select known matches which require currency in exchange and add up the numbers. So currency related intents do NOT come as *primary*, the intents which one considers to fulfill through exchanging something that will fulfill them for that 'crowdfunded' currency act as *primary* ones here.
I use *primary* and *secondary* not as how we would represent particular offer-request or agreement that it leads to. But as what that agent actually wants and has for that agent what you may refer to as *use* value, or will have *use* value for an agent which will eventually act as consumer of some product or service (including work and usage among services).

**Comment by fosterlynn** [comment]
>I use primary and secondary not as how we would represent particular offer-request or agreement that it leads to.

Yes, I understand you philosophically and in the context of how to think about money in society, which is a big subject and can include all kinds of power relations in addition to just facilitating exchange.  And of course accountability in crowdfunding is a good thing.  But for VF, I guess i'm interested in how we represent it to be most helpful to people who want to represent their intents and exchanges.

**Comment by elf-pavlik** [comment]
example of exchange template - apples for peanuts
```yaml
'@context': https://w3id.org/valueflows/v1-dev
id: https://bumblebee.exampe/af5be97e-8f3d-4342-aaa4-e8fcb919df22#EchangeTemplate
type: ExchangeTemplate
author: https://bumblebee.example/#identity
audienceClass: foaf:Agent # anyone but could specify 'members only' etc.
offering:
  id: https://bumblebee.example/a2f50930-0f45-41cf-8c6c-b5ae4266bf6d#IntentTemplate
  type: IntentTemplate
  appliesTo: https://bumblebee.example/7178082f-f41c-4b17-b58a-0cd0931bf423#intent # issue 300kg of apples
  eachQuantity:
    numericValue: 1
    unit: unit:Kilogram
requesting:
  id: https://bumblebee.example/48c52043-1f84-4c92-bb7f-d725697cb97a
  type: IntentTemplate
  appliesTo: https://bumblebee.example/5fb45882-afe3-4eb5-a9c4-99a6f3a32bc4#stock # receive 20kg of peanuts
  eachQuantity:
    numericValue: 0.067
    unit: unit:Kilogram
```

it uses *vf-x:appliesTo* to reference the original **primary intents** https://github.com/valueflows/valueflows/issues/118

variant for apples for MXN

```yaml
'@context': https://w3id.org/valueflows/v1-dev
id: https://bumblebee.exampe/af5be97e-8f3d-4342-aaa4-e8fcb919df22#EchangeTemplate
type: ExchangeTemplate
author: https://bumblebee.example/#identity
audienceClass: foaf:Agent # anyone but could specify 'members only' etc.
offering:
  id: https://bumblebee.example/a2f50930-0f45-41cf-8c6c-b5ae4266bf6d#IntentTemplate
  type: IntentTemplate
  appliesTo: https://bumblebee.example/7178082f-f41c-4b17-b58a-0cd0931bf423#intent # issue 300kg of apples
  eachQuantity:
    numericValue: 1
    unit: unit:Kilogram
requesting:
  id: https://bumblebee.example/a4dec296-fe88-4697-b633-bbe6b8e1b62e#IntentTemplate
  type: IntentTemplate
  eachQuantity:
    numericValue: 25
    unit: iso4217:MXN
```

it could also reference **secondary intent** (receive MXN) with *appliesTo* but as we discussed we don't expect that it will get created

attempt to combine alternatives using *requestingAlternatives*

```yaml
'@context': https://w3id.org/valueflows/v1-dev
id: https://bumblebee.exampe/af5be97e-8f3d-4342-aaa4-e8fcb919df22#EchangeTemplate
type: ExchangeTemplate
author: https://bumblebee.example/#identity
audienceClass: foaf:Agent # anyone but could specify 'members only' etc.
offering:
  id: https://bumblebee.example/a2f50930-0f45-41cf-8c6c-b5ae4266bf6d#IntentTemplate
  type: IntentTemplate
  appliesTo: https://bumblebee.example/7178082f-f41c-4b17-b58a-0cd0931bf423#intent # issue 300kg of apples
  eachQuantity:
    numericValue: 1
    unit: unit:Kilogram
requestingAlternatives:
  - id: https://bumblebee.example/48c52043-1f84-4c92-bb7f-d725697cb97a
    type: IntentTemplate
    appliesTo: https://bumblebee.example/5fb45882-afe3-4eb5-a9c4-99a6f3a32bc4#stock # receive 20kg of peanuts
    eachQuantity:
      numericValue: 0.067
      unit: unit:Kilogram
  - id: https://bumblebee.example/a4dec296-fe88-4697-b633-bbe6b8e1b62e#IntentTemplate
    type: IntentTemplate
    eachQuantity:
      numericValue: 25
      unit: iso4217:MXN
```

*eachQuantity* defines units for that exchange, so we have offered unit for one of requested units
i used *IntentTemplate* since we want them to scale, based on *appliesTo* and history of events and commitments which referenced intent can expose one can calculate 'remaining quantity' for that intent.
